logger.go

```diff
@@ -14,147 +14,157 @@ type Logger struct {
 }
 
 // NewLogger creates a Logger instance with no additional data.
-func NewLogger() Logger {
-	return Logger{
+func NewLogger() *Logger {
+	return &Logger{
 		loggingT: &logging,
 	}
 }
 
 // WithContext creates a logger from a context.Context
-func WithContext(ctx context.Context) Logger {
+func WithContext(ctx context.Context) *Logger {
 	data, _ := ctx.Value(contextKeyData).([]interface{})
 	prefix, _ := ctx.Value(contextKeyPrefix).(string)
-	return Logger{
+	return &Logger{
 		loggingT: &logging,
 		data:     data,
 		prefix:   prefix,
 	}
 }
 
 // WithPrefix creates a Logger with a given prefix.
-func WithPrefix(prefix string) Logger {
-	return Logger{
+func WithPrefix(prefix string) *Logger {
+	return &Logger{
 		loggingT: &logging,
 		prefix:   prefix,
 	}
 }
 
 // WithData creates a Logger with a given set of backend data.
 // All args provided will be wrapped with glog.Data prior to being sent.
-func WithData(args ...interface{}) Logger {
-	return Logger{
+func WithData(args ...interface{}) *Logger {
+	return &Logger{
 		loggingT: &logging,
 		data:     args,
 	}
 }
 
 // WithPrefix creates a Logger from an existing logger with a specified prefix.
 // Any prefix on the input Logger will be replaced.
-func (l Logger) WithPrefix(prefix string) Logger {
-	newLogger := l
-	newLogger.prefix = prefix
-	return newLogger
+func (l *Logger) WithPrefix(prefix string) *Logger {
+	if l == nil {
```

Comment:
Perhaps it l == nil return glog.WithPrefix(...) ?
(same for all the ones below)
It's more of an approach question, I guess. On the one hand, it's a bug. On the other hand, it's one that's very easy to fix... Not sure which one I prefer, just thinking out loud

Reply:
Doing it that way is a bit too "magical" for my liking. If you apply a nil receiver, you're doing something wrong.

Thinking about it, maybe I shouldn't do the check and just let it fail here.